<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Standardized Test Analysis

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

Decide on your problem statement that will guide your analysis for this project. For guidelines, sample prompts, or inspiration, check out the README.

**To-Do:** *Replace this cell with your problem statement.*

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

The SAT and ACT are standardized tests that many colleges and universities in the United States require for their admissions process. This score is used along with other materials such as grade point average (GPA) and essay responses to determine whether or not a potential student will be accepted to the university.

The SAT has two sections of the test: Evidence-Based Reading and Writing and Math ([*source*](https://www.princetonreview.com/college/sat-sections)). The ACT has 4 sections: English, Mathematics, Reading, and Science, with an additional optional writing section ([*source*](https://www.act.org/content/act/en/products-and-services/the-act/scores/understanding-your-scores.html)). They have different score ranges, which you can read more about on their websites or additional outside sources (a quick Google search will help you understand the scores for each test):
* [SAT](https://collegereadiness.collegeboard.org/sat)
* [ACT](https://www.act.org/content/act/en.html)

Standardized tests have long been a controversial topic for students, administrators, and legislators. Since the 1940's, an increasing number of colleges have been using scores from sudents' performances on tests like the SAT and the ACT as a measure for college readiness and aptitude ([*source*](https://www.minotdailynews.com/news/local-news/2017/04/a-brief-history-of-the-sat-and-act/)). Supporters of these tests argue that these scores can be used as an objective measure to determine college admittance. Opponents of these tests claim that these tests are not accurate measures of students potential or ability and serve as an inequitable barrier to entry. Lately, more and more schools are opting to drop the SAT/ACT requirement for their Fall 2021 applications ([*read more about this here*](https://www.cnn.com/2020/04/14/us/coronavirus-colleges-sat-act-test-trnd/index.html)).

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

### Choose your Data

There are 10 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online.

* [`act_2017.csv`](./data/act_2017.csv): 2017 ACT Scores by State
* [`act_2018.csv`](./data/act_2018.csv): 2018 ACT Scores by State
* [`act_2019.csv`](./data/act_2019.csv): 2019 ACT Scores by State
* [`act_2019_ca.csv`](./data/act_2019_ca.csv): 2019 ACT Scores in California by School
* [`sat_2017.csv`](./data/sat_2017.csv): 2017 SAT Scores by State
* [`sat_2018.csv`](./data/sat_2018.csv): 2018 SAT Scores by State
* [`sat_2019.csv`](./data/sat_2019.csv): 2019 SAT Scores by State
* [`sat_2019_by_intended_college_major.csv`](./data/sat_2019_by_intended_college_major.csv): 2019 SAT Scores by Intended College Major
* [`sat_2019_ca.csv`](./data/sat_2019_ca.csv): 2019 SAT Scores in California by School
* [`sat_act_by_college.csv`](./data/sat_act_by_college.csv): Ranges of Accepted ACT & SAT Student Scores by Colleges

**To-Do:** *Fill out this cell with the datasets you will use for your analysis. Write a brief description of the contents for each dataset that you choose.*

### Outside Research

Based on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [20]:
# Code:
def my_mean(events):
    return (sum(events)/len(events))

In [19]:
#Function Test
numbers = [1,2,3,4,5,6,7,8,9,9,8,7,6,5,4,3,2,1]
my_mean(numbers)

5.0

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [17]:
# Code:
def std_dev(events):
    mean = my_mean(events)
    over_n = 1/len(events)
    rms = list(map(lambda event: ((event-mean)**2)*over_n, events))
    return round(sum(rms)**0.5,3)

In [18]:
#Function Test
numbers = (1, 2, 3, 4) 
std_dev(numbers)

1.118

3. Data cleaning function:
    
    Write a function that takes in a string that is a number and a percent symbol (ex. '50%', '30.5%', etc.) and converts this to a float that is the decimal approximation of the percent. For example, inputting '50%' in your function should return 0.5, '30.5%' should return 0.305, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [11]:
# Code:
def calc_decimal(percent_string):
    return float(percent_string[:-1])/100

In [12]:
#Function Test
calc_decimal('35.67%')

0.3567

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [1]:
# Imports:
import numpy as np
import pandas as pd
from random import randint
import matplotlib.pyplot as plt
import seaborn as sns

## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind the minimum & maximum possible values for each test/subtest).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of percents (ex. '50%', '30.5%', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to floats! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [1]:
%pwd

'C:\\Users\\bgrif\\Desktop\\GA_DSI\\projects\\project1_SAT_ACT\\project_1\\code'

In [189]:
#SAT College Board Benchmarks 12th Grade: M-530/ERW-480, 11th Grade: M-510/ERW-970
#SAT converted score: ACT-21, SAT-[1060-1090]
#https://collegereadiness.collegeboard.org/about/scores/benchmarks
%ls ..

 Volume in drive C is Windows
 Volume Serial Number is 72F8-2D4C

 Directory of C:\Users\bgrif\Desktop\GA_DSI\projects\project1_SAT_ACT\project_1

09/19/2020  10:25 PM    <DIR>          .
09/19/2020  10:25 PM    <DIR>          ..
09/19/2020  10:25 PM             1,218 .gitignore
09/20/2020  10:21 PM    <DIR>          code
09/20/2020  09:59 PM    <DIR>          data
09/19/2020  10:25 PM            17,339 README.md
               2 File(s)         18,557 bytes
               4 Dir(s)  321,994,792,960 bytes free


In [58]:
# Code:
#Import Data Sets
actdf = pd.read_csv('../data/act_2019_ca.csv') #2019 ACT Scores in California by School
satdf = pd.read_csv('../data/sat_2019_ca.csv') #2019 SAT Scores in California by School
collegedf = pd.read_csv('../data/sat_act_by_college.csv') #Ranges of Accepted ACT & SAT Student Scores by Colleges
med_incomedf = pd.read_csv('../data/median_household_income_2018.csv') #U.S. Census Bureau
equitydf = pd.read_csv('../data/equitable_county_rank.csv') #2019 California School District Equitablility Ranking

#Display Data
#print(actdf.head())
#print(satdf.head())
#print(collegedf.head())
#print(equitydf.head())

#Drop the empty (last) row
actdf.drop(actdf.tail(1).index,inplace=True)
satdf.drop(satdf.tail(1).index,inplace=True)

#RType == 'C' entries contain aggregate County info
act_countydf = actdf[actdf['RType'] == 'C']
sat_countydf = satdf[satdf['RType'] == 'C']

#RType == 'D' entries contain aggregate School District info
act_districtdf = actdf[actdf['RType'] == 'D']
sat_districtdf = satdf[satdf['RType'] == 'D']

#Create filters for special cases where score records are not available(or don't exist)
#Filter out DISTRICT records where no seniors took exam, but more than 15 were enrolled
null_district_actdf = act_districtdf[act_districtdf['PctGE21'].isnull()==False]
null_district_satdf = sat_districtdf[sat_districtdf['PctBothBenchmark12'].isnull()==False]
#Filter out COUNTY records where no seniors took exam, but more than 15 were enrolled
null_county_actdf = act_countydf[act_countydf['PctGE21'].isnull()==False]
null_county_satdf = sat_countydf[sat_countydf['PctBothBenchmark12'].isnull()==False]

#Filter out DISTRICT records where less than 15 enrolled seniors took exam
district_scores_actdf = null_district_actdf[null_district_actdf['PctGE21']!='*'].copy()
district_scores_satdf = null_district_satdf[null_district_satdf['PctBothBenchmark12']!='*'].copy()
#Filter out COUNTY records where less than 15 enrolled seniors took exam
county_scores_actdf = null_county_actdf[null_county_actdf['PctGE21']!='*'].copy()
county_scores_satdf = null_county_satdf[null_county_satdf['PctBothBenchmark12']!='*'].copy()



#print(act_countydf.groupby('RType')[['NumTstTakr','Enroll12']].agg( ['mean', 'max', 'min', 'std'] ))
#print(act_countydf[['CName','NumGE21','PctGE21']].info());
#print(act_countydf['NumTstTakr'].std())

#Use this code later to look at statistics after adding a column pct_participation = ('NumTstTakr' / 'Enroll12')
#print(act_countydf[['NumTstTakr', 'Enroll12']].agg( ['mean', 'max', 'min', 'std'] ))
#print(act_districtdf.groupby('CName')[['NumTstTakr','Enroll12']].agg( ['mean', 'max', 'min', 'std'] ))

'''for row in actdf.index:
    if actdf['PctGE21'][row].isnull()==False:
        print(actdf[actdf['PctGE21'].isnull()==False])
        print(actdf[actdf['PctGE21']=='*'])'''

#act_scoresdf.dropna(inplace=True)
#act_scores = act_scoresdf[act_scoresdf['PctGE21'] != 'NaN']
#print(act_scoresdf.info())
#sat_scoresdf = 

#Drop comma from median household income column
for row in range(58):
    med_incomedf['median_household_income'] = med_incomedf['median_household_income'][row].replace(',','')
#print(actdf.columns)

#Fix Incorrect Data Types
actdf['CCode'] = actdf['CCode'].apply(int)
satdf['CCode'] = satdf['CCode'].apply(int)

actdf['CDCode'] = actdf['CDCode'].apply(int)
satdf['CDCode'] = satdf['CDCode'].apply(int)

actdf['Enroll12'] = actdf['Enroll12'].apply(int)
satdf['Enroll12'] = satdf['Enroll12'].apply(int)

actdf['NumTstTakr'] = actdf['NumTstTakr'].apply(int)
satdf['NumTSTTakr12'] = satdf['NumTSTTakr12'].apply(int)

#Additional column dtypes to fix for 'PctGE21'-(ACT data) & 'PctBothBenchmark12'-(SAT data)
county_scores_actdf['NumGE21'] = county_scores_actdf['NumGE21'].apply(float)
county_scores_actdf['PctGE21'] = county_scores_actdf['PctGE21'].apply(float)

county_scores_satdf['NumTSTTakr12'] = county_scores_satdf['NumTSTTakr12'].apply(float)
county_scores_satdf['PctBothBenchmark12'] = county_scores_satdf['PctBothBenchmark12'].apply(float)

print(county_scores_actdf[['NumGE21','PctGE21']].agg( ['mean', 'max', 'min', 'std'] ))
print(county_scores_satdf[['NumTSTTakr12','PctBothBenchmark12']].agg( ['mean', 'max', 'min', 'std'] ))

district_scores_actdf['PctGE21'] = district_scores_actdf['PctGE21'].apply(float)
#print(district_scores_actdf.groupby('DName')['PctGE21'].agg(['mean', 'max', 'min', 'std']))
med_incomedf['median_household_income'] = med_incomedf['median_household_income'].apply(int)

#Check df for changes
#print(act_countydf.info())
#print(sat_countydf.info())
print(county_scores_actdf.info())
print(county_scores_satdf.info())

#Remove excess columns
actdf = actdf[['CCode', 'CDCode', 'RType', 'DName', 'CName', 'Enroll12', 'NumTstTakr','NumGE21', 'PctGE21']]
satdf = satdf[['CCode', 'CDCode', 'RType', 'CName','Enroll12', 'NumTSTTakr12', 'TotNumBothBenchmark12', 'PctBothBenchmark12']]

#Rename Columns
actdf = actdf.rename(columns={'CCode':'county_code', 'CDCode':'district_code', 'RType':'record_type', 'DName':'district', 'CName':'county','Enroll12':'enrolled', 'NumTstTakr':'tested', 'NumGE21':'surpassed', 'PctGE21':'pct_surpassed'})
satdf = satdf.rename(columns={'CCode':'county_code', 'CDCode':'district_code', 'RType':'record_type', 'CName':'county','Enroll12':'enrolled', 'NumTSTTakr12':'tested', 'TotNumBothBenchmark12':'surpassed', 'PctBothBenchmark12':'pct_surpassed'})
equitydf = equitydf.rename(columns={'Rank*':'rank', 'School District':'district', 'Score':'score','Expenditures for Public Elementary and Secondary Schools per Pupil':'expenditures_per_pupil','Income by School District':'income'})

#Additional Cleaning
#ID equitable counties from respective districts



#print(satdf.info())
#print(med_incomedf.info())
#print(equitydf.info())


          NumGE21    PctGE21
mean   841.759259  56.757407
max   9182.000000  87.100000
min      8.000000  20.000000
std   1675.299939  17.189004
      NumTSTTakr12  PctBothBenchmark12
mean   3098.054545           49.723091
max   48676.000000           73.180000
min      16.000000           18.920000
std    7238.653654           13.253115
<class 'pandas.core.frame.DataFrame'>
Int64Index: 54 entries, 1064 to 1121
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   CDS          54 non-null     float64
 1   CCode        54 non-null     float64
 2   CDCode       54 non-null     float64
 3   SCode        54 non-null     float64
 4   RType        54 non-null     object 
 5   SName        0 non-null      object 
 6   DName        0 non-null      object 
 7   CName        54 non-null     object 
 8   Enroll12     54 non-null     float64
 9   NumTstTakr   54 non-null     float64
 10  AvgScrRead   54 non-null     object 
 11  A

In [132]:
distcount_dict = {}
for row in actdf.index:
    #print(row)
    c = actdf['county'][row]
    d = actdf['district'][row]
    #print(c)
    #print(d)
    if c in distcount_dict:
        if d not in distcount_dict[c]:
            distcount_dict[c].append(d)
    else:
        #distcount_dict[c] = set([d])
        #distcount_dict.update({c:d}) 
        distcount_dict.update({c:[d]})

In [54]:
distcount_dict.keys()

dict_keys(['Riverside', 'Los Angeles', 'Kern', 'Santa Clara', 'San Francisco', 'San Bernardino', 'Humboldt', 'Contra Costa', 'Tulare', 'Ventura', 'Alameda', 'Monterey', 'Orange', 'Santa Barbara', 'Amador', 'Fresno', 'Napa', 'Sonoma', 'Shasta', 'Mendocino', 'Solano', 'Placer', 'San Benito', 'Santa Cruz', 'San Mateo', 'San Luis Obispo', 'Sacramento', 'San Joaquin', 'Stanislaus', 'Merced', 'San Diego', 'Kings', 'Nevada', 'Lassen', 'Butte', 'Inyo', 'Imperial', 'Calaveras', 'Siskiyou', 'Lake', 'Trinity', 'Sutter', 'Del Norte', 'El Dorado', 'Madera', 'Plumas', 'Mono', 'Colusa', 'Tuolumne', 'Yuba', 'Tehama', 'Yolo', 'Glenn', 'Sierra', 'Mariposa', 'Modoc', 'Marin', 'Alpine', 'State of California'])

In [52]:
distcount_dict.get('Riverside')

['Beaumont Unified',
 'Riverside Unified',
 'Banning Unified',
 'Perris Union High',
 'California School for the Deaf-Riverside (State Special Schl)',
 'Moreno Valley Unified',
 'Palm Springs Unified',
 'Corona-Norco Unified',
 'Temecula Valley Unified',
 'Val Verde Unified',
 'Coachella Valley Unified',
 'Lake Elsinore Unified',
 'Riverside County Office of Education',
 'Hemet Unified',
 'Alvord Unified',
 'Desert Sands Unified',
 'Jurupa Unified',
 'San Jacinto Unified',
 'Murrieta Valley Unified',
 nan,
 'Menifee Union Elementary',
 'Nuview Union ',
 'Palo Verde Unified',
 'Nuview Union']

In [42]:
actdf['county'].sort_values().unique()

NameError: name 'df' is not defined

In [77]:
actdf[actdf['county']=='Riverside']['district'].sort_values()

1391       Alvord Unified
1673       Alvord Unified
711        Alvord Unified
839        Alvord Unified
139       Banning Unified
              ...        
2182    Val Verde Unified
311     Val Verde Unified
1410    Val Verde Unified
1749    Val Verde Unified
1093                  NaN
Name: district, Length: 108, dtype: object


In [27]:
actdf[actdf['county']=='Riverside']

county_code  district_code record_type                  district  \
0              33        3366993           S          Beaumont Unified   
100            33        3367215           S         Riverside Unified   
139            33        3366985           S           Banning Unified   
140            33        3366985           S           Banning Unified   
148            33        3366993           S          Beaumont Unified   
...           ...            ...         ...                       ...   
2222           33        3367124           S     Moreno Valley Unified   
2225           33        3375200           S   Murrieta Valley Unified   
2256           33        3373676           S  Coachella Valley Unified   
2258           33        3367082           S             Hemet Unified   
2259           33        3367082           S             Hemet Unified   

         county  enrolled  tested surpassed pct_surpassed  
0     Riverside        18       0       NaN           NaN  
100   Riverside       434     123        40         32.52  
139   Riverside       218      51        13         25.49  
140   Riverside        91       0       NaN           NaN  
148   Riverside       652      48        23         47.92  
...         ...       ...     ...       ...           ...  
2222  Riverside       450     121        23         19.01  
2225  Riverside       884     165       111         67.27  
2256  Riverside        63       4         *             *  
2258  Riverside       437      72        25         34.72  
2259  Riverside        67      21        16         76.19  

[108 rows x 9 columns]

In [40]:
distcount_dict['Riverside']

'Hemet Unified'

In [173]:
for row in range(len(equitydf['district'])):
    for index in range(len(actdf['district'])):
        if equitydf['district'][row] == actdf['district'][index]:
            equitydf['county'][row] = actdf['county'][index]
            print(equitydf['district'][row]+' found')
            
equitydf['county']

KeyError: 'county'

In [ ]:
print(equitydf['county'])

In [156]:
#for index in range(len(equitydf['district'])):
    #equitydf.loc[ index, [ 'district' ] ]  = equitydf['district'][index].replace('School District','')
#equitydf.loc[ :, [ 'district' ] ]  
#for index in range(len(equitydf['district'])):
    #actdf['district'].str.find(equitydf['district'][index], start=0, end=None).unique()
df = actdf.loc[ :, [ 'county', 'district' ] ].sort_values('district').dropna()
#df.set_index('county')['district'].to_dict()
#actdf['district'].unique()
#actdf.sort_values('district').head(25)
#data_dict = actdf.to_dict()
#districts = list(actdf['district'].unique())
grouped = df.groupby(df['county'])
grouped.count()
#Parameters:
#sub: String or character to be searched in the text value in series
#start: int value, start point of searching. Default is 0 which means from the beginning of string
#end: int value, end point where the search needs to be stopped. Default is None.

district
county                       
Alameda                    87
Amador                      5
Butte                      25
Calaveras                   7
Colusa                      9
Contra Costa               55
Del Norte                   4
El Dorado                  14
Fresno                     84
Glenn                      10
Humboldt                   21
Imperial                   17
Inyo                        8
Kern                       54
Kings                      17
Lake                       12
Lassen                      9
Los Angeles               465
Madera                     17
Marin                      15
Mariposa                    2
Mendocino                  21
Merced                     25
Modoc                       4
Mono                        6
Monterey                   33
Napa                       11
Nevada                     11
Orange                    108
Placer                     30
Plumas                      5
Riverside                 107
Sacramento                 78
San Benito                  4
San Bernardino            120
San Diego                 178
San Francisco              23
San Joaquin                55
San Luis Obispo            23
San Mateo                  34
Santa Barbara              34
Santa Clara                72
Santa Cruz                 24
Shasta                     25
Sierra                      2
Siskiyou                   15
Solano                     22
Sonoma                     45
Stanislaus                 43
State of California         1
Sutter                     20
Tehama                      8
Trinity                     5
Tulare                     45
Tuolumne                   10
Ventura                    43
Yolo                       14
Yuba                       10

### 

In [14]:
print(equitydf.columns)

Index(['Rank*', 'School District', 'Score',
       'Expenditures for Public Elementary and Secondary Schools per Pupil',
       'Income by School District'],
      dtype='object')


### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|ACT/SAT|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in the dataframe.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value 
        - *Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest participation rates for the 2017, 2019, or 2019 SAT and ACT?
        - Which states have the highest and lowest mean total/composite scores for the 2017, 2019, or 2019 SAT and ACT?
        - Do any states with 100% participation on a given test have a rate change year-to-year?
        - Do any states show have >50% participation on *both* tests each year?
        - Which colleges have the highest median SAT and ACT scores for admittance?
        - Which California school districts have the highest and lowest mean test scores?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [220]:
#Code:

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [221]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

Don't forget to create your README!

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!